docs/proposals/container-init.md

```diff
@@ -0,0 +1,333 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Pod initialization
+
+@smarterclayton
+
+March 2016
+
+## Proposal and Motivation
+
+Within a pod there is a need to initialize local data or adapt to the current
+cluster environment that is not easily achieved in the current container model.
+Containers start in parallel after volumes are mounted, leaving no opportunity
+for coordination between containers without specialization of the image. If
+two containers need to share common initialization data, both images must
+be altered to cooperate using filesystem or network semantics, which introduces
+coupling between images. Likewise, if an image requires configuration in order
+to start and that configuration is environment dependent, the image must be
+altered to add the necessary templating or retrieval.
+
+This proposal introduces the concept of an **init container**, one or more
```

Comment:
#831 has evolved to cover 2 features:

1. Image volumes: pull an image and mount it in. Would fill similar role as volumes-from in Docker and most uses of MPM within Google. Would enable file systems that are dynamically composed from multiple images. https://www.youtube.com/watch?v=_uJlTllziPI

2. Container volumes: Execute an image to populate a volume. Maybe could subsume (1). I want to use this approach rather than create more volume sources similar to gitRepo. I could imagine using it to pull from private git repos, GCS, S3, etc.

If we wanted to support in-place updates, and I'm not sure we do, behavior should be identical to updates for downward API, configmap, and secret volumes. If attached to a volume, we could support atomic data updates.

If a container populating an emptyDir fails, it can be retried, by flushing the contents of the dir and creating a new one. That's more complicated with init containers, which presumably wouldn't be required to have a 1-to-1 relationship between container and volume.

This factoring also makes it more straightforward in the API to add an initialization container to a pod together with its corresponding volume.

That's not to say that I think init containers have no utility, but for the purpose of initializing volumes, they impair the system's ability to manage the process.